futures-util/src/compat/compat01as03.rs

```diff
@@ -95,6 +120,168 @@ impl<St: Stream01> Stream03 for Compat01As03<St> {
+                Ok(AsyncSink01::NotReady(i)) => {
+                    self.buffer = Some(i);
+                    task03::Poll::Pending
+                }
+                Err(e) => task03::Poll::Ready(Err(e)),
+            },
+            None => task03::Poll::Ready(Ok(())),
+        }
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Result<(), Self::SinkError>> {
+        let item = self.buffer.take();
+        match self.in_notify(lw, |f| match item {
+            Some(i) => match f.start_send(i) {
+                Ok(AsyncSink01::Ready) => f.poll_complete().map(|i| (i, None)),
+                Ok(AsyncSink01::NotReady(t)) => {
+                    Ok((Async01::NotReady, Some(t)))
+                }
+                Err(e) => Err(e),
+            },
+            None => f.poll_complete().map(|i| (i, None)),
+        }) {
+            Ok((Async01::Ready(_), _)) => task03::Poll::Ready(Ok(())),
+            Ok((Async01::NotReady, item)) => {
+                self.buffer = item;
+                task03::Poll::Pending
+            }
+            Err(e) => task03::Poll::Ready(Err(e)),
+        }
+    }
+
+    fn poll_close(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Result<(), Self::SinkError>> {
+        let item = self.buffer.take();
+        let close_started = self.close_started;
+
+        match self.in_notify(lw, |f| match item {
+            Some(i) => match f.start_send(i) {
+                Ok(AsyncSink01::Ready) => {
+                    match f.poll_complete() {
+                        Ok(Async01::Ready(_)) => {
+                            match <S as Sink01>::close(f) {
+                                Ok(i) => Ok((i, None, true)),
+                                Err(e) => Err(e)
+                            }
+                        },
+                        Ok(Async01::NotReady) => Ok((Async01::NotReady, None, false)),
+                        Err(e) => Err(e)
+                    }
+                },
+                Ok(AsyncSink01::NotReady(t)) => {
+                    Ok((Async01::NotReady, Some(t), close_started))
+                }
+                Err(e) => Err(e),
+            },
+            None => if close_started {
+                match <S as Sink01>::close(f) {
+                    Ok(i) => Ok((i, None, true)),
+                    Err(e) => Err(e)
+                }
+            } else {
+                match f.poll_complete() {
+                    Ok(Async01::Ready(_)) => {
+                        match <S as Sink01>::close(f) {
+                            Ok(i) => Ok((i, None, true)),
+                            Err(e) => Err(e)
+                        }
+                    },
+                    Ok(Async01::NotReady) => Ok((Async01::NotReady, None, close_started)),
+                    Err(e) => Err(e)
+                }
+            },
+        }) {
```

Comment:
This looks correct, but I think using early exits for the first things tried would make this a bit easier to read, a quick (untested) sketch of what I mean:

```rust
let item = self.buffer.take();
let close_started = self.close_started;

let result = self.in_notify(lw, |f| {
    if !close_started {
        if let Some(item) = item {
            if let AsyncSink01::NotReady(item) = f.start_send(item)? {
                return Ok((Async01::NotReady, item, false));
            }
        }

        if let Async01::NotReady = f.poll_complete()? {
            return Ok((Async01::NotReady, None, false));
        }
    }

    Ok((<S as Sink01>::close(f)?, None, true))
});

match result {
    Ok((Async01::Ready(_), _, _)) => task03::Poll::Ready(Ok(())),
    Ok((Async01::NotReady, item, close_started)) => {
        self.buffer = item;
        self.close_started = close_started;
        task03::Poll::Pending
    }
    Err(e) => task03::Poll::Ready(Err(e)),
}
```